guacamole/src/main/java/org/apache/guacamole/rest/directory/DirectoryResource.java

```diff
@@ -386,46 +417,210 @@ public Map<String, ExternalType> getObjects(
+    public APIPatchResponse patchObjects(List<APIPatch<ExternalType>> patches)
             throws GuacamoleException {
 
-        // Apply each operation specified within the patch
-        for (APIPatch<String> patch : patches) {
+        // An outcome for each patch included in the request. This list
+        // may include both success and failure responses, though the
+        // presense of any failure would indicated that the entire
+        // request has failed and no changes have been made.
+        List<APIPatchOutcome> patchOutcomes = new ArrayList<>();
 
-            // Only remove is supported
-            if (patch.getOp() != APIPatch.Operation.remove)
-                throw new GuacamoleUnsupportedException("Only the \"remove\" "
-                        + "operation is supported.");
+        // Perform all requested operations atomically
+        directory.tryAtomically(new AtomicDirectoryOperation<InternalType>() {
 
-            // Retrieve and validate path
-            String path = patch.getPath();
-            if (!path.startsWith("/"))
-                throw new GuacamoleClientException("Patch paths must start with \"/\".");
+            @Override
+            public void executeOperation(boolean atomic, Directory<InternalType> directory)
+                    throws GuacamoleException {
+
+                // If the underlying directory implentation does not support
+                // atomic operations, abort the patch operation. This REST
+                // endpoint requires that operations be performed atomically.
+                if (!atomic)
+                    throw new GuacamoleUnsupportedException(
+                            "Atomic operations are not supported. " +
+                            "The patch cannot be executed.");
+
+                // Keep a list of all objects that have been successfully
+                // added or removed
+                Collection<InternalType> addedObjects = new ArrayList<>();
+                Collection<String> removedIdentifiers = new ArrayList<>();
+
+                // A list of all responses associated with the successful
+                // creation of new objects
+                List<APIPatchOutcome> creationSuccesses = new ArrayList<>();
+
+                // True if any operation in the patch failed. Any failure will
+                // fail the request, though won't result in immediate stoppage
+                // since more errors may yet be uncovered.
+                boolean failed = false;
+
+                // Apply each operation specified within the patch
+                for (APIPatch<ExternalType> patch : patches) {
+
+                    // Retrieve and validate path
+                    String path = patch.getPath();
+                    if (!path.startsWith("/"))
+                        throw new GuacamoleClientException("Patch paths must start with \"/\".");
+
+                    if(patch.getOp() == APIPatch.Operation.add) {
```

Comment:
I hesitate to question this at all, because I know you know what you're doing, but i"ll throw a couple of things out there - feel free to ignore them if you've already considered and have reasons for doing this the way you have it written, here...
* Throughout this section, there are multiple calls to `patch.getOp()`. I don't know what the cost is in terms of doing that versus assigning the value to a variable and checking throughout this section? Maybe it's negligible, and doesn't matter, and it makes sense to keep it this way? Or maybe it occupies more memory than it's worth and the cost of doing it this way is less than going the variable route?
* I know there are only two operations here (three if you count the catch-all at the end), but any reason to go `if/else` rather than a `switch()` statement?

Reply:
I would expect the cost of calling `getOp()` to be negligible considering that it doesn't do any calculation and just returns a value. There wouldn't be any harm in creating a copy of the variable in the local scope either though. 

I suspect that the JDK or JVM might optimize away the cost of the extra stack frame for the `getOp()` call, but I'm not 100% sure about that. I'll change it over to a local variable and then we don't have to worry about it.

As for the switch statement - I mostly just wanted to avoid another layer of indentation, since an `if` adds one level, and 
```
switch(op) {
    case add:
         ... stuff ...
```
adds two.

I'll admit that a switch statement is probably the "right way to do it" though. If people have strong preferences here, I'm happy to change it over.

Reply:
No strong preference, just curious.